tests/Unit/DataStructures/Tensor/Test_Slice.cpp

```diff
@@ -0,0 +1,138 @@
+template <typename VectorType>
+void test_variables_slice() noexcept {
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<size_t> sdist{5, 10};
+
+  const size_t x_extents = sdist(gen);
+  const size_t y_extents = sdist(gen);
+  const size_t z_extents = sdist(gen);
+  Variables<tmpl::list<VariablesTestTags_detail::tensor<VectorType>>> vars{
+      x_extents * y_extents * z_extents};
+  const size_t tensor_size =
+      VariablesTestTags_detail::tensor<VectorType>::type::size();
+  Index<3> extents(x_extents, y_extents, z_extents);
+
+  // Test data_on_slice function by using a predictable data set where each
+  // entry is assigned a value equal to its index
+  for (size_t s = 0; s < vars.size(); ++s) {
+    // clang-tidy: do not use pointer arithmetic
+    vars.data()[s] = s;  // NOLINT
+  }
+  Variables<tmpl::list<VariablesTestTags_detail::tensor<VectorType>>>
+      expected_vars_sliced_in_x(y_extents * z_extents, 0.),
+      expected_vars_sliced_in_y(x_extents * z_extents, 0.),
+      expected_vars_sliced_in_z(x_extents * y_extents, 0.);
+  const size_t x_offset = sdist(gen) % x_extents;
+  const size_t y_offset = sdist(gen) % y_extents;
+  const size_t z_offset = sdist(gen) % z_extents;
+
+  for (size_t s = 0; s < expected_vars_sliced_in_x.size(); ++s) {
+    // clang-tidy: do not use pointer arithmetic
+    expected_vars_sliced_in_x.data()[s] = x_offset + s * x_extents;  // NOLINT
+  }
+  for (size_t i = 0; i < tensor_size; ++i) {
+    for (size_t x = 0; x < x_extents; ++x) {
+      for (size_t z = 0; z < z_extents; ++z) {
+        // clang-tidy: do not use pointer arithmetic
+        expected_vars_sliced_in_y
+            .data()[x + x_extents * (z + z_extents * i)] =  // NOLINT
+            i * extents.product() + x + x_extents * (y_offset + z * y_extents);
+      }
+    }
+  }
+  for (size_t i = 0; i < tensor_size; ++i) {
+    for (size_t x = 0; x < x_extents; ++x) {
+      for (size_t y = 0; y < y_extents; ++y) {
+        // clang-tidy: do not use pointer arithmetic
+        expected_vars_sliced_in_z
+            .data()[x + x_extents * (y + y_extents * i)] =  // NOLINT
+            i * extents.product() + x + x_extents * (y + y_extents * z_offset);
+      }
+    }
+  }
+
+  INFO("Test simple slice");
+  CHECK(data_on_slice(get<VariablesTestTags_detail::tensor<VectorType>>(vars),
+                      extents, 0, x_offset) ==
+        get<VariablesTestTags_detail::tensor<VectorType>>(
+            expected_vars_sliced_in_x));
+  CHECK(data_on_slice(get<VariablesTestTags_detail::tensor<VectorType>>(vars),
+                      extents, 1, y_offset) ==
+        get<VariablesTestTags_detail::tensor<VectorType>>(
+            expected_vars_sliced_in_y));
+  CHECK(data_on_slice(get<VariablesTestTags_detail::tensor<VectorType>>(vars),
+                      extents, 2, z_offset) ==
+        get<VariablesTestTags_detail::tensor<VectorType>>(
+            expected_vars_sliced_in_z));
+
+  INFO("Test slice of a boost::optional<Tensor>");
```

Comment:
Test the cases for the return-by-pointer version where one of the arguments is `boost::none` and the other is not.